polygeist/lib/Dialect/Polygeist/Transforms/FunctionSpecialization.cpp

```diff
@@ -0,0 +1,231 @@
+
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/GPU/IR/GPUDialect.h"
+#include "mlir/Dialect/Polygeist/IR/Ops.h"
+#include "mlir/Dialect/Polygeist/IR/Polygeist.h"
+#include "mlir/Dialect/Polygeist/Utils/TransformUtils.h"
+#include "mlir/Dialect/SYCL/IR/SYCLOps.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "function-specialization"
+
+namespace mlir {
+namespace polygeist {
+#define GEN_PASS_DEF_FUNCTIONSPECIALIZATION
+#include "mlir/Dialect/Polygeist/Transforms/Passes.h.inc"
+} // namespace polygeist
+} // namespace mlir
+
+using namespace mlir;
+using AccessorType = VersionConditionBuilder::AccessorType;
+using AccessorPairType = VersionConditionBuilder::AccessorPairType;
+
+static llvm::cl::opt<unsigned> functionSpecializationAccessorLimit(
+    "function-specialization-accessor-limit", llvm::cl::init(5),
+    llvm::cl::desc(
+        "Maximum number of accessors allowed for function specialization"));
+
+//===----------------------------------------------------------------------===//
+// Helper Functions
+//===----------------------------------------------------------------------===//
+
+/// Returns true if \p type is 'memref<?x!sycl.accessor>', and false otherwise.
+static bool isValidMemRefType(Type type) {
+  auto mt = dyn_cast<MemRefType>(type);
+  bool isMemRefWithExpectedShape =
+      (mt && mt.hasRank() && (mt.getRank() == 1) &&
+       ShapedType::isDynamic(mt.getShape()[0]) && mt.getLayout().isIdentity());
+  if (!isMemRefWithExpectedShape)
+    return false;
+
+  return isa<sycl::AccessorType>(mt.getElementType());
+}
+
+/// Returns true if \p arg is a candidate argument.
+static bool isCandidateArgs(Value arg) {
+  if (!isa<AccessorType>(arg))
+    return false;
+  return isValidMemRefType(arg.getType());
+}
+
+/// Populate \p candArgs with candidate arguments from \p call.
+static void collectCandidateArguments(CallOpInterface call,
+                                      SmallVectorImpl<AccessorType> &candArgs) {
+  assert(candArgs.empty() && "Expecting empty candArgs");
+  for (Value arg : call.getArgOperands())
+    if (isCandidateArgs(arg))
+      candArgs.push_back(cast<AccessorType>(arg));
+}
+
+/// Returns the cloned version of \p func.
+static FunctionOpInterface cloneFunction(FunctionOpInterface func) {
+  ModuleOp module = func->getParentOfType<ModuleOp>();
+  FunctionOpInterface clonedFunc = func.clone();
+  std::string newFnName = (func.getName() + ".specialized").str();
+#ifndef NDEBUG
+  module->walk([newFnName](FunctionOpInterface func) {
+    assert(func.getName() != newFnName &&
+           "Expecting new function name to be unique");
+  });
+#endif // NDEBUG
+  clonedFunc.setName(newFnName);
+  OpBuilder builder(func);
+  builder.insert(clonedFunc);
+  return clonedFunc;
+}
+
+/// Add attribute 'sycl.inner.disjoint' to all candidate arguments of \p func.
+static void specializeFunction(FunctionOpInterface func) {
```

Comment:
Suggest to use a more specific name like:
``` 
/// Set the attribute 'sycl.inner.disjoint' on the function \p func arguments provided by \p args.
static void setInnerDisjointAttribute(FunctionOpInterface func, ArrarRef<unsigned> args) {
  assert(llvm::all_of(args, [&](unsigned arg) { return arg < func.getNumParameters(); } && "...");
```
where args contains the indexes of the arguments to set to the attribute.